### PR TITLE
[Documentation][Cookbook] How to integrate a Payment Gateway as a Plugin?

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

image

### CODE REVIEW

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+In the following example, a new gateway will be configured, which will send payment details to external API.
+
+1. Set up a new plugin
+
+.. code-block:: bash
+
+    $ composer create-project sylius/plugin-skeleton ProjectName
+
+1. The first step in newly created repository would be to create a new Gateway Factory.
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
2. The first step in the newly created repository would be to create a new Gateway Factory.
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 2]
```diff
@@ -0,0 +1,353 @@
+How to integrate a Payment Gateway as a Plugin?
+===============================================
+
+Among all possible customization, new gateway provider is one of the most common choices. Level of complexity of payment processing,
+regional limits, and amount of potential payment providers makes it hard to support all possible cases. A custom payment gateway is
+sometimes the only choice.
+
+In the following example, a new gateway will be configured, which will send payment details to external API.
+
+1. Set up a new plugin
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
1. Set up a new plugin using the plugin skeleton
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]: + link to Plugin development guide in a tip box here?

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 4]
```diff
@@ -0,0 +1,353 @@
+How to integrate a Payment Gateway as a Plugin?
+===============================================
+
+Among all possible customization, new gateway provider is one of the most common choices. Level of complexity of payment processing,
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
Among all possible customizations, new gateway provider is one of the most common choices. 
Payment processing complexity,
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+regional limits, and amount of potential payment providers makes it hard to support all possible cases. A custom payment gateway is
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
regional limits and the amount of potential payment providers makes it hard for Sylius core to keep up with all possible cases. 
A custom payment gateway is
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+                'payum.factory_title' => 'Sylius Payment',
+            ]);
+
+            $config['payum.api'] = function (ArrayObject $config) {
+                return $config['api_key'] ?? null;
+            };
+        }
+    }
+
+And at the end of `config/services.yaml` put this:
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
And at the end of `config/services.yaml` add such a configuration for your gateway:
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+
+1. Set up a new plugin
+
+.. code-block:: bash
+
+    $ composer create-project sylius/plugin-skeleton ProjectName
+
+1. The first step in newly created repository would be to create a new Gateway Factory.
+
+In src/Payum/SyliusPaymentGatewayFactory.php put this content
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
Prepare a gateway factory class in ``src/Payum/SyliusPaymentGatewayFactory.php``:
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+            arguments:
+                - 'Acme\SyliusExamplePlugin\Payum\SyliusPaymentGatewayFactory'
+            tags:
+                -
+                    name: payum.gateway_factory_builder
+                    factory: sylius_payment
+
+2. Next, one should create a configuration form, where authorization (or some additional information, like sandbox mode) can be specified.
+
+In src/Form/Type/SyliusGatewayConfigurationType.php put this content:
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
Create the configuration type in ``src/Form/Type/SyliusGatewayConfigurationType.php``:

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+         */
+        public function buildForm(FormBuilderInterface $builder, array $options): void
+        {
+            $builder
+                ->add('api_key', TextType::class)
+            ;
+        }
+    }
+
+And at the end of `config/services.yaml` put this:
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
And add its configuration to `config/services.yaml`:
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+        app.sylius_payment:
+            class: Payum\Core\Bridge\Symfony\Builder\GatewayFactoryBuilder
+            arguments:
+                - 'Acme\SyliusExamplePlugin\Payum\SyliusPaymentGatewayFactory'
+            tags:
+                -
+                    name: payum.gateway_factory_builder
+                    factory: sylius_payment
+
+2. Next, one should create a configuration form, where authorization (or some additional information, like sandbox mode) can be specified.
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
3. Next, one should create a configuration form, where authorization (or some additional information, like sandbox mode) can be specified.
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+        Acme\SyliusExamplePlugin\Form\Type\SyliusGatewayConfigurationType:
+            tags:
+                -
+                    name: form.type
+                -
+                    name: sylius.gateway_configuration_type
+                    label: Sylius Payment
+                    type: sylius_payment
+
+3. To introduce support for new configuration fields, we need to create a value object which will be passed to action, so
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
4. To introduce support for new configuration fields, we need to create a value object which will be passed to action, so
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+we can use these values.
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]: these values?? which ones?

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+
+In src/Payum/SyliusApi.php
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
In src/Payum/SyliusApi.php
```
You should describe what it is.

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+            $this->apiKey = $apiKey;
+        }
+
+        public function getApiKey(): string
+        {
+            return $this->apiKey;
+        }
+    }
+
+In src/Payum/SyliusPaymentGatewayFactory.php
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]: Describe here what you're doing

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+                'payum.action.status' => new StatusAction(),
+            ]);
+
+            $config['payum.api'] = function (ArrayObject $config) {
+                return new SyliusApi($config['api_key']);
+            };
+        }
+    }
+
+From now on, new Payment Gateway should be available in admin panel.
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
From now on, your new Payment Gateway should be available in the admin panel.
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+
+// New gateway configuration type
+
+4. Configure new payment method in the admin panel
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
5. Configure new payment method in the admin panel
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+
+// New payment method
+
+5. Configure required actions
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
6. Configure required actions
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+
+We will create two actions: CaptureAction and StatusAction. The first one will be responsible for sending to the external system
+payment amount, currency, and api key configured in previously created form, while the second one will translate response HTTP codes to the proper state of payment.
+
+5.1. Create StatusAction and add it to SyliusPaymentGatewayFactory
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
6.1. Create StatusAction and add it to the SyliusPaymentGatewayFactory
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+
+// New gateway configuration type
+
+4. Configure new payment method in the admin panel
+
+// New payment method
+
+5. Configure required actions
+
+We will create two actions: CaptureAction and StatusAction. The first one will be responsible for sending to the external system
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
We will create two actions: CaptureAction and StatusAction. The first one will be responsible for sending data to an external system:
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+payment amount, currency, and api key configured in previously created form, while the second one will translate response HTTP codes to the proper state of payment.
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
payment amount, currency and an API key configured in the previously created form, while the second one will translate response HTTP codes to a proper state of payment.
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+
+5.1. Create StatusAction and add it to SyliusPaymentGatewayFactory
+
+src/Payum/SyliusPaymentGatewayFactory.php
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]: this should be a part of the code block as a comment.

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+            return
+                $request instanceof GetStatusInterface &&
+                $request->getFirstModel() instanceof SyliusPaymentInterface
+            ;
+        }
+    }
+
+StatusAction will update the state of payment based on details provided by CaptureAction. Based on the value of the status code of the HTTP request, the payment status will be adjusted as follows:
+HTTP 400 (Bad request) - payment has failed
+HTTP 200 (OK) - payment succeed
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
HTTP 200 (OK) - payment succeeded
```
?

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+
+5.2. Create CaptureAction as a service
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
6.2. Create a service for handling the CaptureAction
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+        Acme\SyliusExamplePlugin\Payum\Action\CaptureAction:
+            public: true
+            arguments: ['@sylius.http_client']
+            tags:
+                -
+                    name: payum.action
+                    factory: sylius_payment
+                    alias: payum.action.capture
+
+You are ready to make your first checkout with newly created gateway!
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]:
```suggestion
Your shop is ready to handle the first checkout with your newly created gateway!
```

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+
+*On both previously mentioned interceptors, you may configure a status code of the response. Check the behavior of Sylius for code 400 as well!*
+
+For more information, please go to:
+`https://github.com/Payum/Payum/blob/master/docs/index.md`
+`https://docs.sylius.com/en/1.5/book/orders/payments.html`
```
Comment (CoderMaggie) [inline docs/cookbook/payments/custom-payment-gateway.rst]: should be a :doc: link

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+
+    use Symfony\Component\Form\AbstractType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
+    use Symfony\Component\Form\FormBuilderInterface;
+
+    final class SyliusGatewayConfigurationType extends AbstractType
+    {
+        /**
+         * {@inheritdoc}
+         */
```
Comment (GSadee) [inline docs/cookbook/payments/custom-payment-gateway.rst]: Not needed docblock

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 27]
```diff
@@ -0,0 +1,353 @@
...
+        public function buildForm(FormBuilderInterface $builder, array $options): void
+        {
+            $builder
+                ->add('api_key', TextType::class)
```
Comment (GSadee) [inline docs/cookbook/payments/custom-payment-gateway.rst]: Could be in one line

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 28]
```diff
@@ -0,0 +1,353 @@
...
+
+    declare(strict_types=1);
+
+    namespace Acme\SyliusExamplePlugin\Payum;
+
+    final class SyliusApi
+    {
+        /**
+         * @var string
+         */
```
Comment (GSadee) [inline docs/cookbook/payments/custom-payment-gateway.rst]: `/** @var string */`

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 29]
```diff
@@ -0,0 +1,353 @@
...
+src/Payum/SyliusPaymentGatewayFactory.php
+.. code-block:: php
+
+    <?php
+
+    declare(strict_types=1);
+
+    namespace Acme\SyliusExamplePlugin\Payum;
+
+    use Acme\SyliusExamplePlugin\Payum\Action\CaptureAction;
```
Comment (GSadee) [inline docs/cookbook/payments/custom-payment-gateway.rst]: Not used in this class

File: docs/cookbook/payments/custom-payment-gateway.rst [hunk 30]
```diff
@@ -0,0 +1,353 @@
...
+
+*A external request interceptor was used for training purposes. If `https://sylius-payment.free.beeceptor.com` link is not working, create your own on: `https://beeceptor.com/` or `https://ptsv2.com/`*
+
+config/services.yaml
+
+.. code-block:: yaml
+
+        Acme\SyliusExamplePlugin\Payum\Action\CaptureAction:
+            public: true
+            arguments: ['@sylius.http_client']
```
Comment (GSadee) [inline docs/cookbook/payments/custom-payment-gateway.rst]: You are not consistent with defining arguments

**Comment by GSadee** [comment]
The base of this pull-request was changed, you need fetch and reset your local branch 
if you want to add new commits to this pull request. **Reset before you pull, else commits
may become messed-up.**

Unless you added new commits (to this branch) locally that you did not push yet, 
execute `git fetch origin && git reset "payment-gateway-documentation"` to update your local branch.
 
Feel free to ask for assistance when you get stuck :+1:

**Comment by GSadee** [comment]
Thanks, Łukasz! :tada:

**Pull request merged**